dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/DubboShutdownHook.java

```diff
@@ -42,144 +42,133 @@ public class DubboShutdownHook extends Thread {
+    }
 
     /**
-     * Has it already been destroyed or not?
+     * Returns the singleton instance of <code>DubboShutdownHook</code>.
+     *
+     * @return singleton instance
      */
-    private final AtomicBoolean destroyed = new AtomicBoolean(false);
+    public static DubboShutdownHook getInstance() {
+        return SingletonHelper.INSTANCE;
+    }
+
+    private boolean checkExternalManagedModule(ApplicationModel applicationModel) {
+        for (ModuleModel moduleModel : applicationModel.getModuleModels()) {
+            if (moduleModel.isLifeCycleManagedExternally() && !moduleModel.isDestroyed()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected static int getServerShutdownTimeout(ApplicationModel appModel) {
+        try {
+            return ConfigurationUtils.getServerShutdownTimeout(appModel);
+        } catch (RuntimeException re) {
+            return 0;
+        }
+    }
 
     /**
-     * Whether ignore listen on shutdown hook?
+     * Checks whether external managed {@code ModuleModel} exists, and wait until the corresponding external
+     * managed modules of {@code ApplicationModel} has been destroyed or the serverShutdownTimeout expired.
+     *
      */
-    private final boolean ignoreListenShutdownHook;
+    private void checkAndWaitUntilTimeout() {
+        long start = System.currentTimeMillis();
+        /* To avoid shutdown conflicts between Dubbo and Spring,
+         * wait for the modules bound to Spring to be handled by Spring until timeout.
+         */
+        Map<ApplicationModel, Integer> serverShutdownWaits = FrameworkModel.getAllInstances().stream()
+                .flatMap(fwkModel -> fwkModel.getAllApplicationModels().stream())
+                .distinct()
+                .filter(appModel ->
+                        appModel.getModuleModels().stream().anyMatch(ModuleModel::isLifeCycleManagedExternally))
+                .collect(Collectors.toMap(
+                        appModel -> appModel,
+                        DubboShutdownHook::getServerShutdownTimeout,
+                        (existingValue, newValue) -> existingValue,
+                        HashMap::new));
+
+        for (Map.Entry<ApplicationModel, Integer> entry : serverShutdownWaits.entrySet()) {
+            ApplicationModel applicationModel = entry.getKey();
+            Integer val = entry.getValue();
+
+            if (val.intValue() <= 0 || applicationModel.isDestroyed()) {
+                continue;
+            }
 
-    public DubboShutdownHook(ApplicationModel applicationModel) {
-        super("DubboShutdownHook");
-        this.applicationModel = applicationModel;
-        Assert.notNull(this.applicationModel, "ApplicationModel is null");
-        ignoreListenShutdownHook = Boolean.parseBoolean(
-                ConfigurationUtils.getProperty(applicationModel, CommonConstants.IGNORE_LISTEN_SHUTDOWN_HOOK));
```

Comment:
Should support this

Reply:
@AlbumenJ because gracefulshutdown is composed with framework model, then wonder to know what will happen if two app model have different ignore_listen_shutdown configuration inner same framework model?
* Dubbo Application[1.0](DUBBO_INTERNAL_APPLICATION) -- those **internal app** should be skipped ?
* Dubbo Application[1.1](provider1)

Reply:
Yes, we should provide a way to disable `ShutdownHook`. In some special scenario, we cannot depend on it.a

Reply:
Currently, the most valuable ops in the Hook are graceful shutdown command (just for dubbo protocol).